### PR TITLE
Documentation for merge sort algorithm

### LINKED ISSUE #46: Content: Merge Sort
## 📚 Documentation

Add the article for Merge Sort. 

### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

Hi @HarshCasper, can I work on this issue?

### PR DESCRIPTION
**Pull request by ELITA04**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

Added documentation for Merge Sort

### Checklist

- [X] I've read the contribution guidelines.
- [X] I've checked the issue list before deciding what to submit.
- [X] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests
Fixes #46

### CODE REVIEW

**Comment by ELITA04** [comment]
Hi @HarshCasper,

Could you please help me understand why the build is failing? 

Also please review my work and let me know if any changes required.

Thank you!

File: sidebars.js [hunk 1]
```diff
@@ -32,7 +32,8 @@ module.exports = {
       label: 'Sorting Algorithms',
       items: [
         'sorting-algorithms/sorting-algorithms',
-        'sorting-algorithms/bubble-sort'
+        'sorting-algorithms/bubble-sort',
+        '/sorting-algorithms/merge-sort'
```
Comment (HarshCasper) [inline sidebars.js:36]: Your build is failing because of this. Don't keep relative paths. Just keep `sorting-algorithms/merge-sort` here

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Please fix the build and the merge conflicts. I would be only then able to review the article on a live view.

**Comment by ELITA04** [comment]
@HarshCasper,

That was a silly mistake on my part, sorry about it!

I did the changes please review it now.

**Comment by ELITA04** [comment]
@HarshCasper,
I solved the conflict issue as well. Please have a look

Thanks

File: docs/sorting-algorithms/merge-sort.mdx [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+## Explaination
+
+The merge sort breaks an array down into smaller arrays until the size of the array is reduced to 1. Since a single element is always considered to be sorted, two contiguous elements can be merged. The merge sort algorithm therefore breaks the array down into smaller sub-arrays on the way down the recursion tree. On the way back up, it merges these smaller sub-arrays into a single array.
+
+To understand better, lets take an example of an unsorted array
+
+| 12  | 33  | 28  | 10  | 35  | 19  | 42  | 44  |
+| --- | --- | --- | --- | --- | --- | --- | --- |
+
+**Stage 1: Divide**
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:19]: Use Heading-4

File: docs/sorting-algorithms/merge-sort.mdx [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+</td><td>
+
+| 44 |
+| -- | 
+</td></tr> 
+</table>
+
+As we have finally achieved atomic values, this cannot be further divided.
+
+**Stage 2: Compare and Combine**
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:108]: Use Heading-4

File: docs/sorting-algorithms/merge-sort.mdx [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+</td></tr> 
+</table>
+
+And after the final merging, we get the following
+
+| 10  | 12  | 19  | 28  | 33  | 35  | 42  | 44  |
+| --- | --- | --- | --- | --- | --- | --- | --- |
+
+## Alogrithmic Steps
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:158]: Just keep `Algorithm`

File: docs/sorting-algorithms/merge-sort.mdx [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+Assume an array `n`
+- Divide this array into 2 parts of length `n/2` and `n - n/2`. Let us call these arrays left half and right half respectively.
+- Recursively sort the left and right half arrays
+- Merge the left and right array to get a fully sorted array
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:162]: Put full-stops at the sentence end

File: docs/sorting-algorithms/merge-sort.mdx [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+	return merge(left, right)
+}
+```
+
+</TabItem>
+</Tabs>
+
+## Sample Input and Output
+**Input**
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:302]: Keep heading-4

File: docs/sorting-algorithms/merge-sort.mdx [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+```
+Enter an array with spaces
+5 3 2 4 5 84 6 1 54
+```
+
+**Output**
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:308]: Keep heading-4

File: docs/sorting-algorithms/merge-sort.mdx [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+```
+1 2 3 4 5 5 6 54 84
+```
+
+**Input**
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:313]: Keep heading-4

File: docs/sorting-algorithms/merge-sort.mdx [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+```
+Enter an array with spaces
+4 8 7 2 11 1 3
+```
+
+**Output**
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:319]: Keep heading-4

File: docs/sorting-algorithms/merge-sort.mdx [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+```
+1 2 3 4 7 8 11
+```
+
+## Complexity Analysis
+
+`n:` is the number of elements in the array.
+
+- **Time Complexity:** O (n log n)
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:328]: Keep the values in backticks

File: docs/sorting-algorithms/merge-sort.mdx [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+
+- **Space Complexity:** O (n)
```
Comment (HarshCasper) [inline docs/sorting-algorithms/merge-sort.mdx:330]: Keep the values in backticks

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Tiny changes required. Sending it off for a peer review.

File: docs/sorting-algorithms/merge-sort.mdx [hunk 12]
```diff
@@ -0,0 +1,335 @@
+---
+title: Merge Sort
+slug: /sorting-algorithms/merge-sort
+---
+
+## Introduction
+
+Merge Sort is a sorting method that uses the _divide-and-conquer_ paradigm. It splits a data list to be sorted into two halves and then sorts the two half data sets recursively. Finally, these are merged to obtain a complete sorted list.
+
+## Explaination
```
Comment (nidh-ish) [inline docs/sorting-algorithms/merge-sort.mdx:10]: Make it Explanation.

File: docs/sorting-algorithms/merge-sort.mdx [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+
+The merge sort breaks an array down into smaller arrays until the size of the array is reduced to 1. Since a single element is always considered to be sorted, two contiguous elements can be merged. The merge sort algorithm therefore breaks the array down into smaller sub-arrays on the way down the recursion tree. On the way back up, it merges these smaller sub-arrays into a single array.
+
+To understand better, lets take an example of an unsorted array
```
Comment (nidh-ish) [inline docs/sorting-algorithms/merge-sort.mdx:14]: Make it let us.

File: docs/sorting-algorithms/merge-sort.mdx [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+slug: /sorting-algorithms/merge-sort
+---
+
+## Introduction
+
+Merge Sort is a sorting method that uses the _divide-and-conquer_ paradigm. It splits a data list to be sorted into two halves and then sorts the two half data sets recursively. Finally, these are merged to obtain a complete sorted list.
+
+## Explaination
+
+The merge sort breaks an array down into smaller arrays until the size of the array is reduced to 1. Since a single element is always considered to be sorted, two contiguous elements can be merged. The merge sort algorithm therefore breaks the array down into smaller sub-arrays on the way down the recursion tree. On the way back up, it merges these smaller sub-arrays into a single array.
```
Comment (nidh-ish) [inline docs/sorting-algorithms/merge-sort.mdx:12]: Add comma before and after therefore.

File: docs/sorting-algorithms/merge-sort.mdx [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+| -- | -- | 
+
+</td><td>
+
+| 42 | 44 |
+| -- | -- | 
+</td></tr> 
+</table>
+
+In the next iteration, we compare the 2 lists together, and merge them into a list placing all in a sorted order.
```
Comment (nidh-ish) [inline docs/sorting-algorithms/merge-sort.mdx:137]: Make it - placing them all in sorted order.

**Review by nidh-ish — CHANGES_REQUESTED** [review]

**Comment by ELITA04** [comment]
Hi @HarshCasper  and @nidh-ish,

I have made the changes requested.

**Review by nidh-ish — APPROVED** [review]
LGTM.

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**